crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+#![allow(unused)]
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    fmt::Display,
+    sync::Arc,
+};
+
+use miden_objects::{
+    accounts::AccountId,
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+
+use crate::store::TransactionInputs;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchId(u64);
+
+impl Display for BatchId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BatchId {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
+
+impl Display for BlockNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BlockNumber {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+pub struct TransactionPool {
+    /// The latest inflight state of each account.
+    ///
+    /// Accounts without inflight transactions are not stored.
+    account_state: BTreeMap<AccountId, (Digest, TransactionId)>,
```

Comment:
Should the value be a triplet `(Digest, TransactionId, Option<BlockNumber>)`? The block number would be used to track the block at the new state was inserted into the store. This way, we'd be able to easily remove all states for blocks which are too far in the past.

Also, it may make sense to create a newtype for this to expose an interface that is tailored for our needs.

Reply:
Currently I use the transaction ID of the stale block to check this. This has the benefit that the account ID is part of the transaction data so we only have to check the specific state instead of iterating over all of it.